express/component_spec.py

```diff
@@ -0,0 +1,346 @@
+"""This module defines classes to represent an Express component specification."""
+import copy
+import json
+import yaml
+import pkgutil
+import types
+import typing as t
+from pathlib import Path
+from dataclasses import dataclass, asdict
+
+import jsonschema.exceptions
+from jsonschema import Draft4Validator
+from jsonschema.validators import RefResolver
+
+from express.exceptions import InvalidComponentSpec
+from express.io_utils import load_yaml
+from express.common import Field
+from express.pipeline_utils import get_kubeflow_type
+
+
+@dataclass
+class KubeflowInput:
+    """
+    Kubeflow component input argument
+    Attributes:
+        name: name of the argument
+        description: argument description
+        type: the python argument type (str, int, ...)
+    """
+
+    name: str
+    description: str
+    type: str
+
+
+@dataclass
+class KubeflowOutput:
+    """
+    Kubeflow component output argument
+    Attributes:
+        name: name of the argument
+        description: argument description
+    """
+
+    name: str
+    description: str
+
+
+T = t.TypeVar("T", KubeflowInput, KubeflowOutput)
+
+
+@dataclass
+class KubeflowComponent:
+    """
+    A class representing a Kubeflow Pipelines component.
+    Attributes:
+        name: The name of the component.
+        description: The description of the component.
+        image: The Docker image url for the component.
+        args: A dictionary of component arguments.
+        inputs: The input parameters for the component.
+        outputs: The output parameters for the component.
+        command: The command to run the component.
+    """
+
+    name: str
+    description: str
+    image: str
+    args: t.Dict[str, t.Dict[str, str]]
+    inputs: t.Union[KubeflowInput, t.List[KubeflowInput]] = None
+    outputs: t.Union[KubeflowOutput, t.List[KubeflowOutput]] = None
+    command: t.List[t.Union[str, t.Dict[str, str]]] = None
+
+    def __post_init__(self):
+        self.inputs = self._add_defaults_values(
+            self.inputs, self._get_default_inputs(), KubeflowInput
+        )
+
+        self.outputs = self._add_defaults_values(
+            self.outputs, self._get_default_outputs(), KubeflowOutput
+        )
+
+        self._add_input_args()
+        self._set_component_run_command()
+
+    @staticmethod
+    def _get_default_inputs() -> t.List[KubeflowInput]:
+        """
+        Set default inputs to components
+        """
+        inputs = [
+            KubeflowInput(
+                name="input_manifest_path",
+                description="Path to the the input manifest",
+                type="String",
+            ),
+            KubeflowInput(
```

Comment:
Why are these always added? I would expect all of them to be defined in the Fondant component spec.

Reply:
Do you mean in the yaml Component that the user has to define or directly set in the `FondantComponent` class. I have a preference for the latter since the user then does not have to define the manifest inputs and outputs (or need to know about those concepts)

Reply:
yes @NielsRogge pointed this out as well, it is now removed from the default inputs. 

This means that we're now parsing each argument individually as opposed to passing one json string with all the arguments. Kubeflow has a weird way of dealing with bools and iterables so we have to make sure we have a proper way of dealing with this in the `Datasets` class